onnxruntime/core/providers/cuda/math/softmax_blockwise_impl.cuh

```diff
@@ -128,29 +124,30 @@ blockReduce(AccumT* smem, AccumT val,
   return smem[0];
 }
 
-
-template <template<typename, typename> class Reduction, int ILP, typename T, typename AccumT>
+template <template <typename, typename> class Reduction, int ILP, typename T, typename AccumT>
 __device__ __forceinline__ AccumT
 ilpReduce(int shift,
           T* data,
           int size,
           const Reduction<T, AccumT>& r,
-          AccumT defaultVal)
-{
+          AccumT defaultVal) {
   using LoadT = aligned_vector<T, ILP>;
   AccumT threadVal = defaultVal;
   int offset = threadIdx.x;
 
   // shift and do 1
-  if(shift > 0){
+  if (shift > 0) {
     data -= shift;
     size += shift;
-    if(threadIdx.x >= shift){
+    if (threadIdx.x >= shift && threadIdx.x < size) {
```

Comment:
could you pls add the small test that will fail without the fixes into UTs ?

Reply:
Unit test with op level will call softmax_warpwise implementation when using small element size. We can keep it unmerged and wait for next PR about softmax. The next PR will include kernel explorer test which directly call softmax_blockwise with small element size.

Reply:
OK I see, op level UT will not goes to blockwise softmax, so this bug won't be triggered. The bug fix is for tunable op scenario.